File: samples/java/spark/sources/SparkNaiveBayesCSR.java
/* file: SparkNaiveBayesCSR.java */
/*******************************************************************************
* Copyright 2017-2019 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
//  Content:
//      Java sample of Naive Bayes classification in the distributed processing
//      mode.
//
//      The program trains the Naive Bayes model on a supplied training data set
//      and then performs classification of previously unseen data.
////////////////////////////////////////////////////////////////////////////////
*/

package DAAL;

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;

import org.apache.spark.api.java.*;
import org.apache.spark.api.java.function.*;
import org.apache.spark.SparkConf;

import com.intel.daal.algorithms.classifier.prediction.PredictionResult;
import com.intel.daal.algorithms.classifier.prediction.PredictionResultId;
import com.intel.daal.algorithms.classifier.prediction.NumericTableInputId;
import com.intel.daal.algorithms.classifier.prediction.ModelInputId;
import com.intel.daal.algorithms.classifier.training.InputId;
import com.intel.daal.algorithms.classifier.training.TrainingResultId;
import com.intel.daal.algorithms.classifier.training.PartialResultId;

import com.intel.daal.algorithms.multinomial_naive_bayes.*;
import com.intel.daal.algorithms.multinomial_naive_bayes.training.*;
import com.intel.daal.algorithms.multinomial_naive_bayes.prediction.*;

import scala.Tuple2;
import com.intel.daal.data_management.data.*;
import com.intel.daal.services.*;

public class SparkNaiveBayesCSR {
    /* Class containing the algorithm results */
    static class NaiveBayesResult {
        public HomogenNumericTable prediction;
    }

    private static final long nClasses          = 20;
    private static final int  nTestObservations = 2000;

    public static NaiveBayesResult runNaiveBayes(DaalContext context,
                                                 JavaRDD<Tuple2<CSRNumericTable, HomogenNumericTable>> trainDataRDD,
                                                 JavaRDD<Tuple2<CSRNumericTable, HomogenNumericTable>> testDataRDD) {
        JavaRDD<TrainingPartialResult> partsRDD = trainLocal(trainDataRDD);
        TrainingPartialResult finalPartRes = reducePartialResults(partsRDD);
        Model model = trainMaster(context, finalPartRes);

        HomogenNumericTable prediction = testModel(context, testDataRDD, model);

        NaiveBayesResult result = new NaiveBayesResult();
        result.prediction = prediction;
        return result;
    }

    private static JavaRDD<TrainingPartialResult> trainLocal(JavaRDD<Tuple2<CSRNumericTable, HomogenNumericTable>> trainDataRDD) {
        return trainDataRDD.map(new Function<Tuple2<CSRNumericTable, HomogenNumericTable>, TrainingPartialResult>() {
            public TrainingPartialResult call(Tuple2<CSRNumericTable, HomogenNumericTable> tup) {
                DaalContext context = new DaalContext();

                /* Create an algorithm to train the Naive Bayes model on local nodes */
                TrainingDistributedStep1Local algorithm = new TrainingDistributedStep1Local(context, Double.class, TrainingMethod.fastCSR,
                                                                                            nClasses);
                /* Set the input data on local nodes */
                tup._1.unpack(context);
                tup._2.unpack(context);
                algorithm.input.set(InputId.data, tup._1);
                algorithm.input.set(InputId.labels, tup._2);

                /* Train the Naive Bayes model on local nodes */
                TrainingPartialResult pres = algorithm.compute();
                pres.pack();

                context.dispose();
                return pres;
            }
        });
    }

    private static TrainingPartialResult reducePartialResults(JavaRDD<TrainingPartialResult> partsRDD) {
        return partsRDD.reduce(new Function2<TrainingPartialResult, TrainingPartialResult, TrainingPartialResult>() {
            public TrainingPartialResult call(TrainingPartialResult p1, TrainingPartialResult p2) {
                DaalContext context = new DaalContext();

                /* Create an algorithm to compute new partial result from two partial results */
                TrainingDistributedStep2Master algorithm = new TrainingDistributedStep2Master(context, Double.class,
                                                                                              TrainingMethod.fastCSR, nClasses);

                /* Set the partial results recieved from the local nodes */
                p1.unpack(context);
                p2.unpack(context);
                algorithm.input.add(TrainingDistributedInputId.partialModels, p1);
                algorithm.input.add(TrainingDistributedInputId.partialModels, p2);

                /* Compute a new partial result from two partial results */
                TrainingPartialResult pres = algorithm.compute();
                pres.pack();

                context.dispose();
                return pres;
            }
        });
    }

    private static Model trainMaster(DaalContext context, TrainingPartialResult partRes) {

        /* Create an algorithm to train the Naive Bayes model on the master node */
        TrainingDistributedStep2Master algorithm = new TrainingDistributedStep2Master(context, Double.class,
                                                                                      TrainingMethod.fastCSR, nClasses);

        /* Add partial results computed on local nodes to the algorithm on the master node */
        partRes.unpack(context);
        algorithm.input.add(TrainingDistributedInputId.partialModels, partRes);

        /* Train the Naive Bayes model on the master node */
        algorithm.compute();

        /* Finalize computations and retrieve the training results */
        TrainingResult trainingResult = algorithm.finalizeCompute();

        return trainingResult.get(TrainingResultId.model);
    }

    private static HomogenNumericTable testModel(DaalContext context, JavaRDD<Tuple2<CSRNumericTable, HomogenNumericTable>> testData,
                                                 final Model model) {

        /* Create algorithm objects to predict values of the Naive Bayes model with the fastCSR method */
        PredictionBatch algorithm = new PredictionBatch(context, Double.class, PredictionMethod.fastCSR, nClasses);

        /* Pass the test data to the algorithm */
        List<Tuple2<CSRNumericTable, HomogenNumericTable>> parts_List = testData.collect();
        for (Tuple2<CSRNumericTable, HomogenNumericTable> value : parts_List) {
            value._1.unpack(context);
            algorithm.input.set(NumericTableInputId.data, value._1);
        }
        algorithm.input.set(ModelInputId.model, model);

        /* Compute the prediction results */
        PredictionResult predictionResult = algorithm.compute();

        /* Retrieve the results */
        return (HomogenNumericTable)predictionResult.get(PredictionResultId.prediction);
    }
}

File: samples/python/spark/sources/spark_NaiveBayesCSR.py
# file: spark_NaiveBayesCSR.py
#===============================================================================
# Copyright 2017-2019 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#===============================================================================

#
#  Content:
#      Python sample of Naive Bayes classification in the distributed processing
#      mode.
#
#      The program trains the Naive Bayes model on a supplied training data set
#      and then performs classification of previously unseen data.
#
from __future__ import print_function
import os
import sys

from pyspark import SparkContext, SparkConf

from daal import step1Local, step2Master
from daal.algorithms.multinomial_naive_bayes import training, prediction
from daal.algorithms import classifier

from distributed_hdfs_dataset import (
    serializeNumericTable, deserializePartialResult, deserializeCSRNumericTable,
    deserializeNumericTable, getMergedCSRDataAndLabelsRDD
)

utils_folder = os.path.join(os.environ['DAALROOT'], 'examples', 'python', 'source')
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTables

nClasses = 20
nTestObservations = 2000


def runNaiveBayes(trainDataRDD, testDataRDD):
    partsRDD = trainLocal(trainDataRDD)
    model = trainMaster(partsRDD)
    return testModel(testDataRDD, model)


def trainLocal(trainDataRDD):

    def mapper(tup):

        key, tables = tup
        csr_table, homogen_table = tables

        # Create an algorithm to train the Naive Bayes model on local nodes
        algorithm = training.Distributed(step1Local, nClasses, method=training.fastCSR)

        # Set the input data on local nodes
        deserialized_csr_table = deserializeCSRNumericTable(csr_table)
        deserialized_homogen_table = deserializeNumericTable(homogen_table)
        algorithm.input.set(classifier.training.data, deserialized_csr_table)
        algorithm.input.set(classifier.training.labels, deserialized_homogen_table)

        # Train the Naive Bayes model on local nodes
        pres = algorithm.compute()
        serialized_pres = serializeNumericTable(pres)

        return (key, serialized_pres)
    return trainDataRDD.map(mapper)


def trainMaster(partsRDD):

    # Create an algorithm to train the Naive Bayes model on the master node
    algorithm = training.Distributed(step2Master, nClasses, gmethod=training.fastCSR)

    parts_list = partsRDD.collect()

    # Add partial results computed on local nodes to the algorithm on the master node
    for key, value in parts_list:
        deserialized_pres = deserializePartialResult(value, training)
        algorithm.input.add(training.partialModels, deserialized_pres)

    # Train the Naive Bayes model on the master node
    algorithm.compute()

    # Finalize computations and retrieve the training results
    trainingResult = algorithm.finalizeCompute()

    return trainingResult.get(classifier.training.model)


def testModel(testData, model):

    # Create algorithm objects to predict values of the Naive Bayes model with the fastCSR method
    algorithm = prediction.Batch(nClasses, method=prediction.fastCSR)

    # Pass the test data to the algorithm
    parts_list = testData.collect()
    for _, (csr, homogen) in parts_list:
        deserialized_csr = deserializeCSRNumericTable(csr)
        algorithm.input.setTable(classifier.prediction.data, deserialized_csr)

    algorithm.input.setModel(classifier.prediction.model, model)

    # Compute the prediction results
    predictionResult = algorithm.compute()

    # Retrieve the results
    return predictionResult.get(classifier.prediction.prediction)


if __name__ == "__main__":

    # Create SparkContext that loads defaults from the system properties and the classpath and sets the name
    sc = SparkContext(conf=SparkConf().setAppName("Spark Naive Bayes").setMaster('local[4]'))

    trainDataFilesPath = "/Spark/NaiveBayesCSR/data/NaiveBayesCSR_train_?.csv"
    trainDataLabelsFilesPath = "/Spark/NaiveBayesCSR/data/NaiveBayesCSR_train_labels_?.csv"
    testDataFilesPath = "/Spark/NaiveBayesCSR/data/NaiveBayesCSR_test_1.csv"
    testDataLabelsFilesPath = "/Spark/NaiveBayesCSR/data/NaiveBayesCSR_test_labels_1.csv"

    # Read the training data and labels from a specified path
    trainDataAndLabelsRDD = getMergedCSRDataAndLabelsRDD(trainDataFilesPath, trainDataLabelsFilesPath, sc)

    # Read the test data and labels from a specified path
    testDataAndLabelsRDD = getMergedCSRDataAndLabelsRDD(testDataFilesPath, testDataLabelsFilesPath, sc)

    # Compute the results of the Naive Bayes algorithm for dataRDD
    predicted = runNaiveBayes(trainDataAndLabelsRDD, testDataAndLabelsRDD)

    # Print the results
    parts_list = testDataAndLabelsRDD.collect()
    for _, (csr, homogen) in parts_list:
        expected = deserializeNumericTable(homogen)

    # Redirect stdout to a file for correctness verification
    stdout = sys.stdout
    sys.stdout = open('NaiveBayesCSR.out', 'w')

    printNumericTables(expected, predicted, "Ground truth", "Classification results",
                       "NaiveBayes classification results (first 20 observations):", 20, flt64=False)

    # Restore stdout
    sys.stdout = stdout

    sc.stop()
